modules/portal/app/controllers/VinylDNS.scala

```diff
@@ -191,28 +240,39 @@ class VinylDNS @Inject()(
     val grantType = "authorization_code"
     val redirectUri = routes.VinylDNS.oidcCallback.absoluteURL()
 
-    wsClient
-      .url(tokenEndpoint)
-      .post(
-        Map(
-          "client_id" -> Seq(clientId),
-          "grant_type" -> Seq(grantType),
-          "redirect_uri" -> Seq(redirectUri),
-          "scope" -> Seq(oidcScope),
-          "code" -> Seq(code),
-          "client_secret" -> Seq(secret)
+    val tokenResponse =
+      wsClient
+        .url(tokenEndpoint)
+        .post(
+          Map(
+            "client_id" -> Seq(clientId),
+            "grant_type" -> Seq(grantType),
+            "redirect_uri" -> Seq(redirectUri),
+            "scope" -> Seq(oidcScope),
+            "code" -> Seq(code),
+            "client_secret" -> Seq(secret)
+          )
         )
-      )
-      .map { resp =>
-        val response = resp.json.as[OidcResponse]
+        .map { resp =>
+          resp.json.as[OidcResponse]
+        }
 
-        //TODO need to be validating these tokens
-        val idDecoded = new String(
-          ByteVector.fromBase64(response.idToken.split("\\.")(1)).get.toArray,
-          "UTF-8").trim
+    val keyStoreResponse =
+      wsClient
+        .url(keyStoreEndpoint)
+        .get()
+        .map { resp =>
+          resp.json.as[JwksResponse]
+        }
 
-        processLoginOidc(Json.parse(idDecoded).as[OidcUserDetails])
-      }
+    for {
+      token <- tokenResponse
+      jwksKeys <- keyStoreResponse
+      kid = getKidFromToken(token)
+      jwksKey = getJwksKeyFromKid(kid, jwksKeys).get
+      publicKey = jwksKeyToPublicKey(jwksKey)
+      idDecoded = Jwt.decodeRawAll(token.idToken, publicKey, Seq(JwtAlgorithm.RS256)).get
+    } yield processLoginOidc(Json.parse(idDecoded._2).as[OidcUserDetails])
```

Comment:
this `._2` should be okay, `Jwt.decodeRawAll` returns a 3 tuple assuming it succeeds